### PR TITLE
fixes #1003 - create a function/procedure apoc.nodes.collapse(node-list, {config})

### LINKED ISSUE #1003: create a function/procedure apoc.nodes.collapse(node-list, {config})
which does the same as apoc.refactor.mergeNodes just virtually

so that would entail:
- create a single virtual node for this node-list, merge the other properties
- add a virtual count property
- remove all virtual self relationships (could be an option to disable this behavior)
- merge all relationships of the same type/direction to the same entity and perhaps add a count property (could be an option to disable this behavior)
- optionally add another label? `:Collapsed` ?

Anything else @johnymontana ?

### PR DESCRIPTION
**Pull request by AngeloBusato**
Fixes #1003

create a function/procedure apoc.nodes.collapse(node-list, {config})

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:

  - Add procedure to merge node into virtualNode without delete the original Node

With this dataset:

````
CREATE 
(:Person {name:'mike'})-[:LIVES_IN]->(:City{name:'rome'}), 
(:Employee{name:'mike'})-[:WORKS_FOR]->(l:Company{name:'Larus'}), 
(:Person {name:'kate'})-[:LIVES_IN]->(:City{name:'london'}), 
(:Employee{name:'kate'})-[:WORKS_FOR]->(n:Company{name:'Neo'}),
(:Employee{name:'john'})-[:WORKS_FOR]->(l),
(:Employee{name:'jane'})-[:WORKS_FOR]->(n)
````
nodes

Calling the apoc.nodes.collapse procedure:

```
MATCH (p:Person)-[r:LIVES_IN]->(c:City), (e:Employee)-[w:WORKS_FOR]->(m:Company) WITH p,r,c,e,w,m WHERE p.name = e.name 
CALL apoc.nodes.collapse([p,e],{properties:'combine', mergeRels:true, countProperties: true}) yield from, rel, to 
return from, rel, to
```

We have this result:

collapse

### CODE REVIEW

**Comment by jexp** [comment]
@AngeloBusato Conflict needs to rebase

**Comment by AngeloBusato** [comment]
@jexp Rebased

**Comment by jexp** [comment]
more examples, please check

# people of a city as a single node

MATCH (p:Person)-[:LIVES_IN]->(c:City)
WITH c, collect(p) as subgraph
CALL apoc.nodes.collapse(subgraph,{properties:'combine', mergeRels:true, countProperties: true}) yield from, rel, to 


# city with its people as a single node
MATCH (p:Person)-[:LIVES_IN]->(c:City)
WITH c, c + collect(p) as subgraph
CALL apoc.nodes.collapse(subgraph,{properties:'ignore', mergeRels:true, countProperties: true}) yield from, rel, to 

1 node (c with its people) left

(p:Person)-[:KNOWS]->(p:Person)-[:LIVES_IN]->(c:City)

(:City)-[:KNOWS]-(:City)

also add these examples + screenshots to the docs.

**Comment by AngeloBusato** [comment]
@jexp PR Updated

With this dataset we have:
````
CREATE  (p:Person {name: 'John'})-[:LIVES_IN]->(c:City{name:'London'}),
(p1:Person {name: 'Mike'})-[:LIVES_IN]->(c),
(p2:Person {name: 'Kate'})-[:LIVES_IN]->(c),
(p3:Person {name: 'Budd'})-[:LIVES_IN]->(c),
(p4:Person {name: 'Alex'})-[:LIVES_IN]->(c)
````

apoc nodes collapse_1

Calling the apoc.nodes.collapse procedure:

````
MATCH (p:Person)-[:LIVES_IN]->(c:City)
WITH c, collect(p) as subgraph
CALL apoc.nodes.collapse(subgraph,{properties:'combine', mergeRels:true, countProperties: true}) yield from, rel, to
return from, rel, to
````
We have this result:
apoc nodes collapse_2

With this dataset we have:
````
CREATE 
(p:Person {name: 'John'})-[:LIVES_IN]->(c:City{name:'London'}),
(p1:Person {name: 'Mike'})-[:LIVES_IN]->(c),
(p2:Person {name: 'Kate'})-[:LIVES_IN]->(c),
(p3:Person {name: 'Budd'})-[:LIVES_IN]->(c),
(p4:Person {name: 'Alex'})-[:LIVES_IN]->(c),
(p1)-[:KNOWS]->(p),
(p2)-[:KNOWS]->(p1),
(p2)-[:KNOWS]->(p3),
(p4)-[:KNOWS]->(p3)
````
apoc nodes collapse_3

Calling the apoc.nodes.collapse procedure:

````
MATCH (p:Person)-[:LIVES_IN]->(c:City)
WITH c, c + collect(p) as subgraph
CALL apoc.nodes.collapse(subgraph,{properties:'ignore', mergeRels:true, countProperties: true}) yield from, rel, to
return from, rel, to
````
We have this result:
apoc nodes collapse_4

With this dataset we have:
````
CREATE 
(p:Person {name: 'John'})-[:LIVES_IN]->(c:City{name:'London'}),
(p1:Person {name: 'Mike'})-[:LIVES_IN]->(c),
(p2:Person {name: 'Kate'})-[:LIVES_IN]->(c),
(p3:Person {name: 'Budd'})-[:LIVES_IN]->(c),
(p4:Person {name: 'Alex'})-[:LIVES_IN]->(c1:City{name: 'Rome'}),
(p1)-[:KNOWS]->(p),
(p2)-[:KNOWS]->(p1),
(p2)-[:KNOWS]->(p3),
(p4)-[:KNOWS]->(p3)
````
apoc nodes collapse_5

Calling the apoc.nodes.collapse procedure:

````
MATCH (p:Person)-[:LIVES_IN]->(c:City)
WITH c, c + collect(p) as subgraph
CALL apoc.nodes.collapse(subgraph,{properties:'ignore', mergeRels:true, countProperties: true}) yield from, rel, to
return from, rel, to
````
We have this result:
apoc nodes collapse_6

**Comment by jexp** [comment]
Cool, thanks a lot.

File: docs/overview.adoc [hunk 1]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
 | apoc.any.property(node/rel/map) | returns property for virtual and real, nodes, rels and maps
 | apoc.create.uuid() | returns a UUID string
 |===
 
+On `apoc.nodes.collapse` with config properties you can choose from 3 different behavior:
+ * "properties": "overwrite" : if there is the same property in more node, in the new one will have the last relationship's/node's property value
+ * "properties": "discard" : if there is the same property in more node, the new one will have the first relationship's/node's property value
+ * "properties": "combine" : if there is the same property in more node, the new one a value's array with all relationship's/node's values
+
+Relationships properties are managed with the same nodes' method, if properties parameter isn't set relationships properties are combined.
```
Comment (jexp) [inline docs/overview.adoc]: combined should **not** be the default as it leads to gigantic array properties.

File: docs/overview.adoc [hunk 2]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
 | apoc.rel.id(rel) | returns id for (virtual) relationships
 | apoc.rel.type(rel) | returns type for (virtual) relationships
 | apoc.any.properties(node/rel/map, [[keys]]) | returns properties for virtual and real, nodes, rels and maps, optionally restrict via keys
 | apoc.any.property(node/rel/map) | returns property for virtual and real, nodes, rels and maps
 | apoc.create.uuid() | returns a UUID string
 |===
 
+On `apoc.nodes.collapse` with config properties you can choose from 3 different behavior:
+ * "properties": "overwrite" : if there is the same property in more node, in the new one will have the last relationship's/node's property value
+ * "properties": "discard" : if there is the same property in more node, the new one will have the first relationship's/node's property value
```
Comment (jexp) [inline docs/overview.adoc]: this should be the default.

File: docs/overview.adoc [hunk 3]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
+ * "properties": "combine" : if there is the same property in more node, the new one a value's array with all relationship's/node's values
+
+Relationships properties are managed with the same nodes' method, if properties parameter isn't set relationships properties are combined.
+
+On `apoc.nodes.collapse` there are many config param:
+ * "mergeRels: true/false" : give the possibility to merge relationships with same type and direction.
+ * "selfRel: true/false" : give the possibility to create the self relationship.
```
Comment (jexp) [inline docs/overview.adoc]: default to false

File: docs/overview.adoc [hunk 4]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
 
+On `apoc.nodes.collapse` with config properties you can choose from 3 different behavior:
+ * "properties": "overwrite" : if there is the same property in more node, in the new one will have the last relationship's/node's property value
+ * "properties": "discard" : if there is the same property in more node, the new one will have the first relationship's/node's property value
+ * "properties": "combine" : if there is the same property in more node, the new one a value's array with all relationship's/node's values
+
+Relationships properties are managed with the same nodes' method, if properties parameter isn't set relationships properties are combined.
+
+On `apoc.nodes.collapse` there are many config param:
+ * "mergeRels: true/false" : give the possibility to merge relationships with same type and direction.
```
Comment (jexp) [inline docs/overview.adoc]: default to true

File: docs/overview.adoc [hunk 5]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
+ * "selfRel: true/false" : give the possibility to create the self relationship.
+ * "countProperties: true/false" : give the possibility count all the properties merged for nodes and relationships.
```
Comment (jexp) [inline docs/overview.adoc]: default to true, it should actually not count properties but number of nodes / relationships merged!!!!

File: docs/overview.adoc [hunk 6]
```diff
@@ -953,13 +953,58 @@ Sometimes type information gets lost, these functions help you to coerce an "Any
...
+
+With this dataset we have:
+
+image::{img}/apoc.nodes.collapse_3.png[width=600]
+
+call the apoc as:
+----
+MATCH (p:Person)-[:LIVES_IN]->(c:City)
+WITH c, c + collect(p) as subgraph
+CALL apoc.nodes.collapse(subgraph,{properties:'ignore', mergeRels:true, countProperties: true}) yield from, rel, to
```
Comment (jexp) [inline docs/overview.adoc]: "ignore" is not a documented option, see above. there is only discard or overwrite.

File: src/main/java/apoc/create/Create.java [hunk 7]
```diff
@@ -149,6 +146,10 @@
     @UserFunction("apoc.create.vNode")
     @Description("apoc.create.vNode(['Label'], {key:value,...}) returns a virtual node")
     public Node vNodeFunction(@Name("label") List<String> labelNames, @Name(value = "props",defaultValue = "{}") Map<String, Object> props) {
+        if (labelNames == null) {
+            labelNames = new ArrayList<>();
+        }
+        labelNames.add("VirtualNode");
```
[the remark this answers is not in the document]
Comment (jexp) [reply src/main/java/apoc/create/Create.java]: make it an option to add an extra label and have the user specify it.

File: src/main/java/apoc/nodes/Nodes.java [hunk 8]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
         }
     }
 
+    @Procedure(mode = Mode.WRITE)
+    @Description("apoc.nodes.collapse([nodes...],[{properties:'overwrite' or 'discard' or 'combine'}]) yield from, rel, to merge nodes onto first in list")
+    public Stream<VirtualPathResult> collapse(@Name("nodes") List<Node> nodes, @Name(value = "config", defaultValue = "") Map<String, Object> config) {
+        if (nodes == null || nodes.isEmpty() || nodes.size() < 2) return Stream.empty();
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: empty is included in < 2
and why empty stream with one node?
Shouldn't we then have a virtual node for one entry?

File: src/main/java/apoc/nodes/Nodes.java [hunk 9]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+                    .map(relationship -> new VirtualPathResult(relationship.getStartNode(), relationship, relationship.getEndNode()));
+        } else {
+            return Stream.of(new VirtualPathResult(first, null, null));
+        }
+    }
+
+    private VirtualNode createVirtualNode(Set<Node> nodes, RefactorConfig conf) {
+        Create create = new Create();
+        Node first = nodes.iterator().next();
+        List<String> labels = StreamSupport.stream(first.getLabels().spliterator(), false).map(Label::name).collect(Collectors.toList());
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: there is a util function

File: src/main/java/apoc/nodes/Nodes.java [hunk 10]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+        labels.add("Collapsed");
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: Only if configured by user.

File: src/main/java/apoc/nodes/Nodes.java [hunk 11]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+        VirtualNode virtualNode = (VirtualNode) create.vNodeFunction(labels, first.getAllProperties());
+        createVirtualRelationship(nodes, virtualNode, first, conf);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: createVirtualRelationships (Plural)

File: src/main/java/apoc/nodes/Nodes.java [hunk 12]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+            node.getLabels().forEach(label -> { // Set all Label
+                if (!virtualNode.hasLabel(label)) {
+                    virtualNode.addLabel(label);
+                }
+            });
+
+            Map<String, Object> properties = node.getAllProperties();
+            PropertiesManager.mergeProperties(properties, virtualNode, conf); // Set all properties
+            if (properties.size() > 0 && conf.isCountProperties()) {
+                virtualNode.setProperty("count", (Integer) virtualNode.getProperty("count", virtualNode.getAllProperties().size()) + properties.size()); // Count all properties merged or added
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: that was a misunderstanding, it should only count the number of nodes merged!!

File: src/main/java/apoc/nodes/Nodes.java [hunk 13]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+            }
+            createVirtualRelationship(nodes, virtualNode, node, conf);
+        });
+        mergeRelationship(conf, virtualNode);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: why do we merge after the fact, not during?

File: src/main/java/apoc/nodes/Nodes.java [hunk 14]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+                }
+            } else {
+                if (startNode.getId() == node.getId()) {
+                    virtualRelationship = virtualNode.createRelationshipTo(endNode, relationship.getType());
+                } else {
+                    virtualRelationship = virtualNode.createRelationshipFrom(startNode, relationship.getType());
+                }
+            }
+            if (virtualRelationship != null) {
+                copyProperties(relationship, virtualRelationship);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: didn't pass config??

File: src/main/java/apoc/nodes/Nodes.java [hunk 15]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+            }
+        });
+    }
+
+    private void mergeRelationship(RefactorConfig conf, VirtualNode virtualNode) {
+        if (conf.getMergeRelsAllowed()) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: shouldn't we merge immediately when creating relationships?

File: src/main/java/apoc/nodes/Nodes.java [hunk 16]
```diff
@@ -136,6 +139,80 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+            if(!conf.hasProperties()) {
+                Map<String, Object> map = Collections.singletonMap("properties", "combine");
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: combine should **not** be the default.

File: src/main/java/apoc/refactor/util/RefactorConfig.java [hunk 17]
```diff
@@ -32,7 +35,9 @@ public RefactorConfig(Map<String,Object> config) {
 			this.propertiesManagement = (Map<String,String>)value;
 		}
 
-		this.mergeRelsAllowed = config.get("mergeRels");
+		this.mergeRelsAllowed = toBoolean(config.get("mergeRels"));
+		this.selfRel = toBoolean(config.get("selfRel"));
+		this.countProperties = toBoolean(config.get("countProperties"));
```
Comment (jexp) [inline src/main/java/apoc/refactor/util/RefactorConfig.java]: this was meant to mean "a property that holds the node/relationship count that was merged"

File: src/main/java/apoc/result/VirtualNode.java [hunk 18]
```diff
@@ -136,6 +136,12 @@ public VirtualRelationship createRelationshipTo(Node node, RelationshipType rela
         return rel;
     }
 
+    public VirtualRelationship createRelationshipFrom(Node start, RelationshipType relationshipType) {
```
Comment (jexp) [inline src/main/java/apoc/result/VirtualNode.java]: Why? is this really needed? it's not part of the Node interface.

File: src/main/java/apoc/create/Create.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (conker84) [reply src/main/java/apoc/create/Create.java]: > make it an option to add an extra label and have the user specify it.

we'll revert it and create a new issue to manage this use-case

File: src/main/java/apoc/result/VirtualNode.java (hunk 18 above)
Comment (conker84) [reply src/main/java/apoc/result/VirtualNode.java]: Unfortunately, we can't create a relationship from a "real" node to a virtual

Comment (conker84) [reply src/main/java/apoc/result/VirtualNode.java]: The only way is to create it a reverse way

File: src/main/java/apoc/create/Create.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jexp) [reply src/main/java/apoc/create/Create.java]: Oki.

**Comment by AngeloBusato** [comment]
@jexp PR updated

File: src/main/java/apoc/nodes/Nodes.java [hunk 19]
```diff
@@ -136,6 +139,94 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+        Node first = nodes.iterator().next();
+        List<String> labels = Util.labelStrings(first);
+        if (conf.isCollapsedLabel()) {
+            labels.add("Collapsed");
+        }
+        VirtualNode virtualNode = (VirtualNode) create.vNodeFunction(labels, first.getAllProperties());
+        createVirtualRelationships(nodes, virtualNode, first, conf);
+        nodes.stream().skip(1).forEach(node -> {
+            node.getLabels().forEach(label -> { // Set all Label
+                if (!virtualNode.hasLabel(label)) {
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: We should move that into the VirtualNode class. To make sure each label is only added once.

File: src/main/java/apoc/nodes/Nodes.java [hunk 20]
```diff
@@ -136,6 +139,94 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+                    createOrMergeVirtualRelationship(virtualNode, refactorConfig, relationship, endNode,  Direction.OUTGOING);
+                } else {
+                    createOrMergeVirtualRelationship(virtualNode, refactorConfig, relationship, startNode,  Direction.INCOMING);
+                }
+            }
+        });
+    }
+
+    private void createOrMergeVirtualRelationship(VirtualNode virtualNode, RefactorConfig refactorConfig, Relationship source, Node node, Direction direction) {
+        Iterable<Relationship> iterator = virtualNode.getRelationships(source.getType(), direction);
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: it's an iterable.

File: src/main/java/apoc/nodes/Nodes.java [hunk 21]
```diff
@@ -136,6 +139,94 @@ public boolean connected(@Name("start") Node start, @Name("start") Node end, @Na
...
+        Optional<Relationship> first = StreamSupport.stream(iterator.spliterator(), false).filter(relationship -> relationship.getEndNode().equals(node)).findFirst();
```
Comment (jexp) [inline src/main/java/apoc/nodes/Nodes.java]: getEndNode() is wrong for direction INCOMING (or BOTH) you need to use rel.getOtherNode(virtualNode)
also make sure there is a test for both incoming and outgoing rels

File: src/main/java/apoc/refactor/util/RefactorUtil.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class RefactorUtil {
+
+    public static void mergeRelsWithSameTypeAndDirectionInMergeNodes(Node node, RefactorConfig config, Direction dir) {
+        for (RelationshipType type : node.getRelationshipTypes()) {
+            StreamSupport.stream(node.getRelationships(dir,type).spliterator(), false)
+                    .collect(Collectors.groupingBy(rel -> Pair.of(rel.getStartNode(), rel.getEndNode())))
```
Comment (jexp) [inline src/main/java/apoc/refactor/util/RefactorUtil.java]: should we look here for direction related to start node?

**Pull request merged**